keps/sig-cli/2775-kubectl-delete-interactivity-delay/README.md

```diff
@@ -0,0 +1,579 @@
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+When a namespace is deleted it also deletes all of the resources under it. The deletion runs without further confirmation, and can be devastating if accidentally run against the wrong namespace (e.g. thanks to hasty tab completion use).  
+
+```  
+kubectl delete namespace prod-backup  
+```
+
+When all namespaces are deleted essentially all resources are deleted. This deletion is trivial to do with the `--all` flag, and it also runs without further confirmation. It can effectively wipe out a whole cluster.  
+
+```
+kubectl delete namespace --all
+```  
+
+There are certainly things cluster operators should be doing to help prevent this user error (like locking down permissions) but it doesn't matter how many mistakes have to pile up to create a perfect storm of a bad thing because we're allowing a bad thing to happen without a confirmation gate.
+
+This KEP proposes two protections when deleting resources with kubectl.
+
+## Motivation
+
+Over the years we've heard stories from users who've accidentally deleted resources in their clusters - often all of their resources. This trend seems to be rising lately as newer folks venture into the Kubernetes/DevOps/Infra world. Experienced users are also not accident proof and have shared tragic tales.
+
+### Goals
+
+* Mitigate the potential for accidental imperative deletes
+
+### Non-Goals
+
+* Server side solutions
+
+## Proposal
+
+1. `kubectl delete [--all | --all-namespaces]` will warn about the destructive action that will be performed and artificially delay for x seconds allowing users a chance to abort.
+
+2. Add a new `--interactive | -i` flag to `kubectl delete` that will require confirmation before deleting resources imperatively (i.e. not `-f`). This flag will be false by default. If this flag is explicitly set to false it will not prompt or artificially delay.
```

Comment:
how does this interact with other consumers of stdin in kubectl (e.g. [client-go credential plugins that make use of stdin](https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/541-external-credential-providers), or manifest reading from stdin like `-f -`)

Reply:
I think it will matter where in the call stack we read stdin. The client-go credential plugins probably happen early the flow? I see this living near the end right inside the delete command.

Will need to play around with it with. We can have some integration tests for these.

What would you like to see about it in the KEP?

Reply:
> The client-go credential plugins probably happen early the flow?

I think it depends on where the client config is constructed.

> What would you like to see about it in the KEP?

Describing specific places we use stdin in conjunction with `kubectl delete` calls today (`-f -`, exec plugins, etc), and that this `--interactive` option cannot be used in combination with those (and exercising the behavior when both are attempted with a testcase)